src/client/opamArg.ml

```diff
@@ -782,62 +779,465 @@ let opamlist_columns =
+
+  val cli_from: OpamCLIVersion.t -> validity
+  val cli_between:
+    OpamCLIVersion.t -> ?replaced:string -> OpamCLIVersion.t -> validity
+  val cli_original: validity
+
+  val cli2_0: OpamCLIVersion.t
+  val cli2_1: OpamCLIVersion.t
+  val cli2_2: OpamCLIVersion.t
+
+  val mk_flag:
+    cli:OpamCLIVersion.t -> validity -> ?section:string -> string list ->
+    string -> bool Term.t
+
+  val mk_opt:
+    cli:OpamCLIVersion.t -> validity -> ?section:string -> ?vopt:'a ->
+    string list -> string -> string -> 'a Arg.converter -> 'a -> 'a Term.t
+
+  val mk_opt_all:
+    cli:OpamCLIVersion.t -> validity -> ?section:string -> ?vopt:'a ->
+    ?default:'a list -> string list -> string -> string -> 'a Arg.converter ->
+    'a list Term.t
+
+  val mk_vflag:
+    cli:OpamCLIVersion.t -> ?section:string -> 'a ->
+    (validity * 'a * string list * string) list -> 'a Term.t
+
+  val mk_vflag_all:
+    cli:OpamCLIVersion.t -> ?section:string -> ?default:'a list ->
+    (validity * 'a * string list * string) list -> 'a list Term.t
+
+  val mk_tristate_opt:
+    cli:OpamCLIVersion.t -> validity -> ?section:string -> string list ->
+    string -> string -> [> `Always | `Auto | `Never ] option Term.t
+
+  type 'a subcommand = validity * string * 'a * string list * string
+  type 'a subcommands = 'a subcommand list
+
+  val mk_subcommands:
+    cli:OpamCLIVersion.t -> 'a subcommands ->
+    'a option Term.t * string list Term.t
+
+  val mk_subcommands_with_default:
+    cli:OpamCLIVersion.t -> 'a default subcommands ->
+    'a option Term.t * string list Term.t
+
+  val bad_subcommand:
+    cli:OpamCLIVersion.t -> 'a default subcommands ->
+    (string * 'a option * string list) -> 'b Term.ret
+
+  val mk_subdoc :
+    cli:OpamCLIVersion.t -> ?defaults:(string * string) list ->
+    'a subcommands -> Manpage.block list
+
+  type command = unit Term.t * Term.info
+
+  val mk_command:
+    OpamCLIVersion.t -> validity -> string -> doc:string ->
+    man:Manpage.block list -> (unit -> unit) Term.t -> command
+
+  val mk_command_ret:
+    OpamCLIVersion.t -> validity -> string -> doc:string ->
+    man:Manpage.block list -> (unit -> unit Term.ret) Term.t -> command
+
+end = struct
+
+  let cli2_0 = OpamCLIVersion.of_string "2.0"
+  let cli2_1 = OpamCLIVersion.of_string "2.1"
+  let cli2_2 = OpamCLIVersion.of_string "2.2"
+
+  type 'b validity_and_content = {
+    valid: OpamCLIVersion.t;
+    removed: (OpamCLIVersion.t * string option) option;
+    content: 'b;
+  }
+
+  type 'a contented_validity =
+    [ `Valid of 'a | `Removed of 'a ] validity_and_content
+
+  type validity = unit validity_and_content
+
+  let elem_of_vr = function `Valid e | `Removed e -> e
+
+  let contented_validity (validity:validity) content : 'a contented_validity =
+    match validity.removed with
+    | None -> { validity with content = `Valid content}
+    | Some _ -> { validity with content = `Removed content}
+
+  let cli_from valid = { valid ; removed = None; content = () }
```

Comment:
This can also be implemented for a more simple use as
```
val cli_from: string -> valitidy
let cli_from version = 
  try
    let valid = OpamCLIVersion.of_string version in
    if OpamCLIVersion.is_supported valid then
      { valid; removed = None; content = () }
    else
      OpamConsole.error_and_exit `Internal_error "not supported"
  with Failure _ -> OpamConsole.error_and_exit `Internal_error "not a valid cli"
```
but it will require to define newer version as supported (for example when we know that we want to deprecate an option in a future opam cli version), and we loose a strong check at compile time replaced by a runtime failure.